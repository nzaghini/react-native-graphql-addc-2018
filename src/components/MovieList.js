import React, { Component } from 'react';
import { ScrollView, Text, ActivityIndicator } from 'react-native';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import { MovieItem } from './';

class MovieList extends Component {

    renderMovies({ allMovies }) {
        // Bad performance for large list. Example only.
        return allMovies.map(movie => {
            return <MovieItem key={movie.id} movie={movie} />;
        });
    }

    render() {
        return (
            <Query query={ALL_MOVIES_QUERY}>
                {({ loading, error, data }) => {
                    if (loading) return <ActivityIndicator style={styles.loadingIndicator} />;
                    if (error) return <Text>{`Error: ${error}`}</Text>;
                    return (
                        <ScrollView>
                            {this.renderMovies(data)}
                        </ScrollView>
                    );
                }}
            </Query>
        );
    }
}

// es6 template string to crete the query
const ALL_MOVIES_QUERY = gql` 
    {
        allMovies{
            id
            title
            year
            director{
                firstName
                lastName
            }
        }
    }
`;

const styles = {
    loadingIndicator: {
        flex: 1,
    },
};

export default MovieList;
