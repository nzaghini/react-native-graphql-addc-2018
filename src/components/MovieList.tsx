import React from "react";
import { ScrollView, Text, ActivityIndicator, StyleSheet } from "react-native";
import { Query } from "react-apollo";
import { MovieItem } from ".";
import { ALL_MOVIES_QUERY } from "../queries/queries.graphql";
import { ALL_MOVIES_QUERY as AllMovies } from "../queries/models/ALL_MOVIES_QUERY";

class MovieList extends React.Component {

    render() {
        return (
            <Query query={ALL_MOVIES_QUERY}>
                {({ loading, error, data }) => {
                    
                    if (loading) {
                        return <ActivityIndicator style={styles.loadingIndicator} />;
                    }
                    if (error) {
                        return <Text>{`Error: ${error}`}</Text>;
                    }
                    return (
                        <ScrollView>
                            {this.renderMovies(data)}
                        </ScrollView>
                    );
                }}
            </Query>
        );
    }

    private renderMovies(allMovies: AllMovies) {
        if (!allMovies || !allMovies.movies) {
            return null;
        }
        return allMovies.movies.map((movie) => {
            if (!movie) { return; }
            return <MovieItem key={movie.id} movie={movie} />;
        });
    }
}

// es6 template string to crete the query
const styles = StyleSheet.create({
    loadingIndicator: {
        flex: 1,
    },
});

export default MovieList;
