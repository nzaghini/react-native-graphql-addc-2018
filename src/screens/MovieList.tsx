import React from "react";
import { ScrollView, Text, ActivityIndicator, StyleSheet } from "react-native";
import { Query } from "react-apollo";
import { MovieItem } from "../components";
import { ALL_MOVIES_QUERY } from "../queries/queries.graphql";
import { ALL_MOVIES_QUERY as AllMovies, 
         ALL_MOVIES_QUERY_movies as Movie } from "../queries/models/ALL_MOVIES_QUERY";
import { ScreenProps } from "../common/ScreenProps";
import { Screens } from "../common/AppNavigator";
import {Â i18n } from "../../i18n/I18n";

class MovieList extends React.Component<ScreenProps> {

    static navigationOptions = {
        title: i18n("movieList.title"),
    };

    constructor(props: ScreenProps) {
        super(props);
    }

    render() {
        return (
            <Query query={ALL_MOVIES_QUERY}>
                {({ loading, error, data }) => {
                    if (loading) {
                        return <ActivityIndicator style={styles.loadingIndicator} />;
                    }
                    if (error) {
                        return <Text>{`Error: ${error}`}</Text>;
                    }
                    return (
                        <ScrollView>
                            {this.renderMovies(data)}
                        </ScrollView>
                    );
                }}
            </Query>
        );
    }

    private navigateTo = (movie: Movie) => {
        this.props.navigation.navigate(Screens.MovieDetails, { movie });
    }

    private renderMovies(allMovies: AllMovies) {
        if (!allMovies || !allMovies.movies) {
            return null;
        }
        return allMovies.movies.map((movie) => {
            return <MovieItem key={movie.id} movie={movie} action={this.navigateTo}/>;
        });
    }
}

// es6 template string to crete the query
const styles = StyleSheet.create({
    loadingIndicator: {
        flex: 1,
    },
});

export default MovieList;
