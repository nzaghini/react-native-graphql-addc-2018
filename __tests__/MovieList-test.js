import React from 'react';
import { MockedProvider } from 'react-apollo/test-utils';
import renderer from 'react-test-renderer';
import gql from 'graphql-tag';
import wait from 'waait';
import MovieList from '../src/components/MovieList';

const ALL_MOVIES_QUERY = gql` 
    {
        allMovies{
            id
            title
            year
            director{
                firstName
                lastName
            }
        }
    }
`;

const mocks = [
    {
        request: {
            query: ALL_MOVIES_QUERY,
        },
        result: {
            data: {
                allMovies: [
                    {
                        id: '1',
                        title: 'Interstellar',
                        year: '2014',
                        director: {
                            firstName: 'test',
                            lastName: 'another'
                        }
                    }
                ],
            },
        },
    },
];

test('renders correctly the loading state', () => {
    console.log(mocks);
    const tree = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <MovieList />
      </MockedProvider>
    );
    
    // Not waiting for the provider to return, so loading is true
    expect(tree.toJSON()).toMatchSnapshot();
});

//** Learned the from https://dev-blog.apollodata.com/testing-apollos-query-component-d575dc642e04 */
test('renders correctly the list of movies', async () => {
    console.log(mocks);
    const tree = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <MovieList />
      </MockedProvider>
    );
    
    await wait(0); // wait for response

    const firstMovieTitle = mocks[0].result.data.allMovies[0].title;
    expect(JSON.stringify(tree)).toContain(firstMovieTitle);
    expect(tree.toJSON()).toMatchSnapshot();
});
