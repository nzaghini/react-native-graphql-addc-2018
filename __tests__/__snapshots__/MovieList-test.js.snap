// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly the list of movies 1`] = `
<RCTScrollView>
  <View>
    <View
      style={
        Object {
          "borderBottomWidth": 1,
          "borderColor": "#ddd",
          "height": 60,
          "justifyContent": "center",
          "paddingLeft": 10,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 18,
          }
        }
      >
        Interstellar
         - 
        2014
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 12,
          }
        }
      >
        test
         
        another
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`renders correctly the loading state 1`] = `
<ActivityIndicator
  animating={true}
  color="#999999"
  hidesWhenStopped={true}
  size="small"
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;
